### PR TITLE
App category page

### PR DESCRIPTION
**Pull request by DwinaTech**
Added category page and each category has articles related to it and every article has button to go to singleArticle page.

### CODE REVIEW

**Comment by DwinaTech** [comment]
[APP / Category page](https://trello.com/c/C9OGW9hC/41-app-category-page)

File: client/src/components/Categories/SingleCategory.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        }
+    }
+    componentDidMount() {
+        apiClient.getArticles()
+            .then(({ data }) => {
+                this.setState({
+                    articles: data
+                })
+            })
+            .catch((err) => { })
```
Comment (kabaros) [inline client/src/components/Categories/SingleCategory.js]: never have an empty catch. Remove it - it's better if the application crashes on an error, so that we fix it, than it goes unnoticed.

File: client/src/components/Categories/SingleCategory.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    render() {
+        const { categoryId } = this.props.match.params;
+        return (
+            <div className="SingleCategory">
+                {this.state.articles.map(article => {
+                    if (article.category._id === categoryId) {
+                        return <div>
+                            <h4>{article.title}</h4>
+                            <a className="btn btn-info btn-sm" href={"/articles/" + article._id + " "}>More Info</a>
```
Comment (kabaros) [inline client/src/components/Categories/SingleCategory.js]: use es6 string interpolation with the backticks ` (instead of + string)

File: client/src/containers/SingleArticle/SingleArticle.js [hunk 3]
```diff
@@ -1,27 +1,37 @@
 import React, { Component } from 'react';
-import apiClient from '../../helpers/apiClient';
+import apiClient from '../../helpers/apiClient'
```
Comment (kabaros) [inline client/src/containers/SingleArticle/SingleArticle.js]: why did you change this file?? - it should have been working (it call getSingleArticle and return a single article not an array)

Please put it back as it was.

File: client/src/components/Categories/CategoriesOverview/CategoriesOverview.js [hunk 4]
```diff
@@ -23,7 +23,7 @@ class CategoriesOverview extends Component {
                 {
                     this.state.categoriesList.map((category) => {
                         return (
-                            <CategoryCard category={category} />
+                            <CategoryCard category={category} href={"/categories/"+ category._id +""}/>
```
Comment (kabaros) [inline client/src/components/Categories/CategoriesOverview/CategoriesOverview.js]: use es6 string interpolation with the backticks ` (instead of + string)

**Review by kabaros — CHANGES_REQUESTED** [review]

File: client/src/containers/SingleArticle/SingleArticle.js (hunk 3 above)
Comment (DwinaTech) [reply client/src/containers/SingleArticle/SingleArticle.js]: these two are the same, so I just left one of them

Comment (DwinaTech) [reply client/src/containers/SingleArticle/SingleArticle.js]: it is alright I thing I have got it

File: client/src/components/Categories/SingleCategory.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    constructor() {
+
+        super();
+        this.state = {
+            articles: []
+        }
+    }
+    componentDidMount() {
+        apiClient.getArticles()
```
Comment (kabaros) [inline client/src/components/Categories/SingleCategory.js]: This works but it is inefficient. Right now, you're getting all articles and then filtering them on the client-side. This is expensive operation and unnecessary.

Looking at changing `/articles` API - so that you can call it saying `/articles?categoryId=_id`, then you can get the category Id on the server from `request.params` and query the database directly by `{category: categoryID}`

Does that make sense?

**Review by kabaros — CHANGES_REQUESTED** [review]
it is good, but you can improve it. Please read my comment and tell me if it's not clear.

File: client/src/components/Categories/SingleCategory.js (hunk 5 above)
Comment (DwinaTech) [reply client/src/components/Categories/SingleCategory.js]: is the same have done with singleArticle?

**Comment by kabaros** [comment]
Similar but it would be getSingleCategory or
getArticlesByCategory(categoryid)

sent from my phone

File: client/src/components/Categories/SingleCategory.js (hunk 5 above)
Comment (DwinaTech) [reply client/src/components/Categories/SingleCategory.js]: Do I need to change one api/articles route in the server side?

Comment (DwinaTech) [reply client/src/components/Categories/SingleCategory.js]: I lost actually

Comment (kabaros) [reply client/src/components/Categories/SingleCategory.js]: you can `/api/articles` so that it reads `req.params` (which will contain {cateogry: _id} if you call `/api/articles?category=_id`  .. then it can pass that object to Mongo which will filter by category.

Comment (kabaros) [reply client/src/components/Categories/SingleCategory.js]: maybe it is `req.queryString` actually not req.params - test and see

Comment (kabaros) [reply client/src/components/Categories/SingleCategory.js]: First step: update`/api/articles` so that it handles a query string (i.e. you can call it with `/api/articles?category=_some_category_id` and it will return only the articles in the category.

You can use req.queryString or req.params - to get an object with the stuff in the queryString (i.e. anything after the question mark)

Comment (DwinaTech) [reply client/src/components/Categories/SingleCategory.js]: sorry for replay late just because I am on the college I am doing coding when I am free, so alright I will try to do what you mentioned.

Comment (DwinaTech) [reply client/src/components/Categories/SingleCategory.js]: That is did not work but I will carry on to understand how is it working.

**Review by kabaros — APPROVED** [review]
